cuda/solver/gmres_kernels.cu

```diff
@@ -48,35 +53,377 @@ namespace cuda {
+
+    divide(exec, next_krylov_basis_norm.get(), next_krylov_basis);
+    // next_krylov_basis /= hessenberg(iter, iter + 1)
+    // End of arnoldi
+}
+
+
+namespace kernel {
+
+
+template <typename ValueType>
+__device__ void calculate_sin_and_cos(const size_type num_cols,
+                                      const ValueType *hessenberg_iter,
+                                      ValueType *givens_sin,
+                                      ValueType *givens_cos,
+                                      const size_type iter)
+{
+    const auto local_id = thread::get_local_thread_id<cuda_config::warp_size>();
+
+    if (hessenberg_iter[iter * num_cols + local_id] == zero<ValueType>()) {
+        givens_cos[iter * num_cols + local_id] = zero<ValueType>();
+        givens_sin[iter * num_cols + local_id] = one<ValueType>();
+    } else {
+        auto hypotenuse =
+            sqrt(hessenberg_iter[iter * num_cols + local_id] *
+                     hessenberg_iter[iter * num_cols + local_id] +
+                 hessenberg_iter[(iter + 1) * num_cols + local_id] *
+                     hessenberg_iter[(iter + 1) * num_cols + local_id]);
+        givens_cos[iter * num_cols + local_id] =
+            abs(hessenberg_iter[iter * num_cols + local_id]) / hypotenuse;
+        givens_sin[iter * num_cols + local_id] =
+            givens_cos[iter * num_cols + local_id] *
+            hessenberg_iter[(iter + 1) * num_cols + local_id] /
+            hessenberg_iter[iter * num_cols + local_id];
+    }
+}
+
+
+template <typename ValueType>
+__device__ void calculate_residual_norm(
+    const size_type num_cols, const ValueType *givens_sin,
+    const ValueType *givens_cos, ValueType *residual_norm,
+    ValueType *residual_norms, const ValueType *b_norm, const size_type iter)
+{
+    const auto local_id = thread::get_local_thread_id<cuda_config::warp_size>();
+
+    residual_norms[(iter + 1) * num_cols + local_id] =
+        -givens_sin[iter * num_cols + local_id] *
+        residual_norms[iter * num_cols + local_id];
+    residual_norms[iter * num_cols + local_id] =
+        givens_cos[iter * num_cols + local_id] *
+        residual_norms[iter * num_cols + local_id];
+    residual_norm[local_id] =
+        abs(residual_norms[(iter + 1) * num_cols + local_id]) /
+        b_norm[local_id];
+}
+
+
+template <size_type block_size, typename ValueType>
+__global__ __launch_bounds__(block_size) void givens_rotation(
+    const size_type num_rows, const size_type num_cols,
+    ValueType *__restrict__ hessenberg_iter, ValueType *__restrict__ givens_sin,
+    ValueType *__restrict__ givens_cos, ValueType *__restrict__ residual_norm,
+    ValueType *__restrict__ residual_norms,
+    const ValueType *__restrict__ b_norm, const size_type iter,
+    const stopping_status *__restrict__ stop_status)
+{
+    const auto local_id = thread::get_local_thread_id<cuda_config::warp_size>();
+    __shared__ UninitializedArray<ValueType, block_size> tmp;
```

Comment:
Why do you need this to be a `__shared__` array? This can be a simple variable because you only read and write to `tmp[local_id]` in any case (as far as I can see).